### PR TITLE
crimson/os/seastore/omap: optimize merge policy

### PR DESCRIPTION
**Pull request by aisakaki**
all unittest passed.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h [hunk 1]
```diff
@@ -660,6 +660,11 @@ class StringKVInnerNodeLayout {
   bool is_overflow(size_t ksize) const {
     return free_space() < (sizeof(omap_inner_key_le_t) + ksize);
   }
+
+  bool is_overflow(const StringKVInnerNodeLayout &rhs) const {
+    return free_space() >= rhs.used_space();
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h:665]: should be free_space() <  rhs.used_space(); ?

File: src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h [hunk 2]
```diff
@@ -1267,6 +1272,11 @@ class StringKVLeafNodeLayout {
   bool is_overflow(size_t ksize, size_t vsize) const {
     return free_space() < (sizeof(omap_leaf_key_le_t) + ksize + vsize);
   }
+
+  bool is_overflow(const StringKVLeafNodeLayout &rhs) const {
+    return free_space() >= rhs.used_space();
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h:1277]: same here.

Comment (aisakaki) [reply src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h]: all addressed.

**Review by liu-chunmei — APPROVED** [review]

File: src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc [hunk 3]
```diff
@@ -370,38 +370,44 @@ OMapInnerNode::merge_entry(
     if (donor->extent_is_below_min()) {
       DEBUGT("make_full_merge l {} r {}", oc.t, *l, *r);
       assert(entry->extent_is_below_min());
-      return l->make_full_merge(oc, r).si_then([liter=liter, riter=riter,
-                                                  l=l, r=r, oc, this] (auto &&replacement){
+      return l->make_full_merge(oc, r
+      ).si_then([liter=liter, riter=riter, l=l, r=r, oc, this](auto &&replacement)
+      {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc:375]: no new line for a single `{`.

If the line is too long, probably:
```
).si_then([liter=liter, riter=riter, l=l, r=r, oc, this]
          (auto &&replacement) {
```

File: src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc [hunk 4]
```diff
@@ -370,38 +370,44 @@ OMapInnerNode::merge_entry(
...
 	LOG_PREFIX(OMapInnerNode::merge_entry);
 	DEBUGT("to update parent: {}", oc.t, *this);
-        journal_inner_update(liter, replacement->get_laddr(), maybe_get_delta_buffer());
+        journal_inner_update(liter, replacement->get_laddr(),
+	  maybe_get_delta_buffer());
```
Comment (cyx1231st) [inline src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc:379]: nit, all argument share the same indention:
```
journal_inner_update(
  liter,
  replacement->get_laddr(),
  maybe_get_delta_buffer());
```
Same below.

**Review by cyx1231st — APPROVED** [review]
LGTM except code style suggestions.

File: src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc (hunk 4 above)
Comment (aisakaki) [reply src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc:381]: all addressed.

**Comment by aisakaki** [comment]
jenkins test make check

**Comment by aisakaki** [comment]
retest this please

**Comment by cyx1231st** [comment]
jenkins test api

**Pull request merged**